### PR TITLE
test case to show can't calculate grad for embedding (and associated fix)

### PR DESCRIPTION
**Pull request by silky**
i've noticed that i don't seem to be able to calculate the gradient w.r.t an embedding, this is the error:

```
ERROR: TensorFlowException TF_INVALID_ARGUMENT "Expected begin and size arguments to be 1-D tensors of size 1, but got shapes [] and [] instead.

 [[Node: Slice_54 = Slice[Index=DT_INT32, T=DT_INT32, _device=\"/job:localhost/replica:0/task:0/cpu:0\"](Shape_52, Const_18, Neg_53)]]"
```

the equivalent python code to the following test works (but i note that in python i can *also* calculate the gradient w.r.t say the variable named below called `op` that is a `Tensor Value` not a `Tensor Ref`; i.e. it's a type error)

i'm looking for a bit of guidance here. i'm fairly sure i should be able to calculate the gradient like this. but i could be convinced/shown that i'm doing it wrong, and there's a different way to get the embedding vars to gradient application ...

interested in thoughts.

if we do keep this test, i should first make a bit of a test lib and put some funtions in there (namely the `assertAllClose`...)

(sidenote: i'm having a truly terrible time with shape instances/Int32s/etc, so i'll one day try and figure out how to clean all that up; maybe i'm just missing some imports or some philosophical viewpoint)

### CODE REVIEW

**Comment by blackgnezdo** [comment]
@fkm3 could you take a look if you have some time?

File: tensorflow-ops/tests/EmbeddingOpsTest.hs [hunk 1]
```diff
@@ -85,6 +91,45 @@ testEmbeddingLookupHasRightShape = testCase "testEmbeddingLookupHasRightShape" $
     values @=? V.fromList [ 1, 1, 1, 0, 0, 0 ]
 
 
+-- | Check that we can calculate gradients w.r.t gradients.
```
Comment (blackgnezdo) [inline tensorflow-ops/tests/EmbeddingOpsTest.hs]: The second gradients presumably should be "embeddings".

**Comment by blackgnezdo** [comment]
@silky you are not missing anything anything deep around shapes and int32s. It's an artifact of TF having started migrating to int64. On Haskell side we should just move to int64 as soon there's enough infrastructure in C++. We have no legacy to worry about and types will help us find the code in need of a change.

@vrv, do you happen to know if int64 shape support is complete enough now to transition?

**Comment by vrv** [comment]
No, you should be cc'ed on the internal bug that describes why this is hard. :/

Though I don't fully understand why it's relevant here: is it that the system can't handle being able to output tensors whose elements are either int32 or int64?

**Comment by blackgnezdo** [comment]
@vrv, it can, we are just forced to choose Int32 or Int64. And I don't want to default to Int32 (which is the thing that works fully, right?) as the long term default will be Int64.

**Comment by vrv** [comment]
Only thing I can suggest is that we coordinate the switch of the default from Int32 to Int64 if/when it happens.

**Comment by fkm3** [comment]
The gradient code is using rank 0 tensors as parameters to `slice`, which is apparently illegal. I remember it working before...

I think we just need to update all of the calls to `CoreOps.slice` in https://github.com/tensorflow/haskell/blob/master/tensorflow-ops/src/TensorFlow/Gradient.hs

**Comment by silky** [comment]
thanks @blackgnezdo, i've added that in.

**Comment by blackgnezdo** [comment]
@silky, did you get to try @fkm3's suggestion yet?

**Comment by silky** [comment]
@blackgnezdo  - nope, i might have some time to try it out this weekend but i'm not sure; i'll report back when i have a go

**Comment by silky** [comment]
@fkm3 @blackgnezdo  - yeah, that basically fixed it; i also needed to convert the last argument to a scalar. it works now!

please review, and i'm happy to create that test lib at some point in next day or so then merge.

**Comment by silky** [comment]
i've added that lib now; let me know if anything else needs to be done!

File: tensorflow-ops/tests/EmbeddingOpsTest.hs [hunk 2]
```diff
@@ -85,6 +92,36 @@ testEmbeddingLookupHasRightShape = testCase "testEmbeddingLookupHasRightShape" $
     values @=? V.fromList [ 1, 1, 1, 0, 0, 0 ]
 
 
+-- | Check that we can calculate gradients w.r.t embeddings.
+testEmbeddingLookupGradients = testCase "testEmbeddingLookupGradients" $ do
+    let xVals = V.fromList ([20, 20] :: [Float]) -- Agrees with "embedding", so gradient should be zero.
```
Comment (blackgnezdo) [inline tensorflow-ops/tests/EmbeddingOpsTest.hs]: We try to keep lines under < 80 chars. Could you check if this line fits?

File: tensorflow-ops/tests/EmbeddingOpsTest.hs [hunk 3]
```diff
@@ -85,6 +92,36 @@ testEmbeddingLookupHasRightShape = testCase "testEmbeddingLookupHasRightShape" $
...
+
+    gs <- TF.runSession $ do
+        grads <- TF.build $ do
+            let shape         = TF.Shape [2, 1] 
+            let embeddingInit = [1, 20] :: [Float]
+            let idValues      = [1, 1]   :: [Int32]
+            let ids           = TF.constant (TF.Shape [1, 2]) idValues
+
+            x <- TF.placeholder (TF.Shape [2]) 
+            embedding <- TF.initializedVariable =<< (TF.render $ TF.constant shape embeddingInit)
```
Comment (blackgnezdo) [inline tensorflow-ops/tests/EmbeddingOpsTest.hs]: TF.render (TF.constant shape embeddingInit)  -- should work too?

File: tensorflow-ops/tests/EmbeddingOpsTest.hs [hunk 4]
```diff
@@ -85,6 +92,36 @@ testEmbeddingLookupHasRightShape = testCase "testEmbeddingLookupHasRightShape" $
...
+-- | Check that we can calculate gradients w.r.t embeddings.
+testEmbeddingLookupGradients = testCase "testEmbeddingLookupGradients" $ do
+    let xVals = V.fromList ([20, 20] :: [Float]) -- Agrees with "embedding", so gradient should be zero.
+    let shape = TF.Shape [2]
+
+    gs <- TF.runSession $ do
+        grads <- TF.build $ do
+            let shape         = TF.Shape [2, 1] 
+            let embeddingInit = [1, 20] :: [Float]
+            let idValues      = [1, 1]   :: [Int32]
```
Comment (blackgnezdo) [inline tensorflow-ops/tests/EmbeddingOpsTest.hs]: Remove extra spaces before ::?

File: tensorflow-ops/tests/EmbeddingOpsTest.hs [hunk 5]
```diff
@@ -85,6 +92,36 @@ testEmbeddingLookupHasRightShape = testCase "testEmbeddingLookupHasRightShape" $
...
+
+            grad <- fmap head (TF.gradients loss [embedding])
+
+            return $ \xs -> TF.runWithFeeds [TF.feed x xs] grad
+
+        grads (TF.encodeTensorData shape xVals :: TF.TensorData Float)
+ 
+    -- Gradients should be zero (or close)
+    assertAllClose gs (V.fromList ([0, 0] :: [Float]))
+
```
Comment (blackgnezdo) [inline tensorflow-ops/tests/EmbeddingOpsTest.hs]: Remove a couple of empty lines?

File: tensorflow-ops/tests/EmbeddingOpsTest.hs [hunk 6]
```diff
@@ -85,6 +92,36 @@ testEmbeddingLookupHasRightShape = testCase "testEmbeddingLookupHasRightShape" $
...
+            let loss = TF.mean (CoreOps.square (TF.abs (op - x))) (TF.scalar (0 :: Int32))
+
+            grad <- fmap head (TF.gradients loss [embedding])
+
+            return $ \xs -> TF.runWithFeeds [TF.feed x xs] grad
+
+        grads (TF.encodeTensorData shape xVals :: TF.TensorData Float)
+ 
+    -- Gradients should be zero (or close)
+    assertAllClose gs (V.fromList ([0, 0] :: [Float]))
```
Comment (blackgnezdo) [inline tensorflow-ops/tests/EmbeddingOpsTest.hs]: [0, 0 :: Float] should work too.

File: tensorflow-ops/src/TensorFlow/Ops.hs [hunk 7]
```diff
@@ -256,6 +257,15 @@ constant (Shape shape') values
                       [def & TensorShape.size .~ x | x <- shape']
                 & tensorVal .~ values
 
+-- | Reshape a N-D tensor down to a scalar.
+toScalar :: (TensorType a)
```
Comment (blackgnezdo) [inline tensorflow-ops/src/TensorFlow/Ops.hs]: This should probably be named scalarize or something similar (extractScalar?). We already have "unScalar".

File: tensorflow-ops/src/TensorFlow/Ops.hs [hunk 8]
```diff
@@ -256,6 +257,15 @@ constant (Shape shape') values
                       [def & TensorShape.size .~ x | x <- shape']
                 & tensorVal .~ values
 
+-- | Reshape a N-D tensor down to a scalar.
+toScalar :: (TensorType a) 
+         => Tensor v a 
+         -> Tensor Value a
+toScalar = flip CoreOps.reshape (vector scalarShape)
```
Comment (blackgnezdo) [inline tensorflow-ops/src/TensorFlow/Ops.hs]: I'd avoid flip here. 
toScalar x = CoreOps.reshape x (vector scalarShape)
Is both shorter and easier to understand.

File: tensorflow-ops/src/TensorFlow/Ops.hs [hunk 9]
```diff
@@ -256,6 +257,15 @@ constant (Shape shape') values
                       [def & TensorShape.size .~ x | x <- shape']
                 & tensorVal .~ values
 
+-- | Reshape a N-D tensor down to a scalar.
```
Comment (blackgnezdo) [inline tensorflow-ops/src/TensorFlow/Ops.hs]: Does it fail at runtime if the tensor has more than 1 element?

File: tensorflow-ops/src/TensorFlow/Ops.hs [hunk 10]
```diff
@@ -256,6 +257,15 @@ constant (Shape shape') values
...
+toScalar :: (TensorType a) 
+         => Tensor v a 
+         -> Tensor Value a
+toScalar = flip CoreOps.reshape (vector scalarShape)
+    where
+        scalarShape = [] :: [Int32]
```
Comment (blackgnezdo) [inline tensorflow-ops/src/TensorFlow/Ops.hs]: Maybe link to [reshape documentation](https://www.tensorflow.org/versions/r0.11/api_docs/python/array_ops.html#reshape)
which has an example?

File: tensorflow-ops/src/TensorFlow/Ops.hs (hunk 7 above)
Comment (silky) [inline tensorflow-ops/src/TensorFlow/Ops.hs]: ah; squeeze, i forgot about it, good suggestion!

Comment (silky) [inline tensorflow-ops/src/TensorFlow/Ops.hs]: i've decided i'm torn

ideally one would have a `rank` associated to Tensors, and we could say that `scalarize` goes from `rank n -> rank 0`, whereas `squeeze` would go from `rank n -> rank m` with `m <= n`. and it might be a nice dream to one day have functions that demand `rank 0` tensors.

maybe that's a bit too demanding at this point though.

File: tensorflow-test/tensorflow-test.cabal [hunk 11]
```diff
@@ -0,0 +1,25 @@
+name:                tensorflow-test
+version:             0.1.0.0
+synopsis:            Friendly layer around TensorFlow bindings.
```
Comment (blackgnezdo) [inline tensorflow-test/tensorflow-test.cabal]: Adjust to match intent?

File: tensorflow-ops/src/TensorFlow/Gradient.hs [hunk 12]
```diff
@@ -402,6 +403,21 @@ type GradientFunc a = NodeDef
 toT :: Output -> Tensor Value a
 toT = Tensor ValueKind
 
+
+-- | Wrapper around `TensorFlow.GenOps.Core.slice` that builds vectors from scalars for
+-- simple slicing operations.
+flatSlice :: forall v1 t i . (TensorType t, OneOf '[Int32, Int64] i)
```
Comment (blackgnezdo) [inline tensorflow-ops/src/TensorFlow/Gradient.hs]: You seem to only need the Int32 version, let's reduce generality and improve call sites?

File: tensorflow-ops/tests/EmbeddingOpsTest.hs [hunk 13]
```diff
@@ -23,32 +23,40 @@ import Data.List (genericLength)
...
 
+
 -- | Tries to perform a simple embedding lookup, with two partitions.
-testEmbeddingLookupHasRightShapeWithPartition = testCase "testEmbeddingLookupHasRightShapeWithPartition" $ do
+testEmbeddingLookupHasRightShapeWithPartition = 
+        testCase "testEmbeddingLookupHasRightShapeWithPartition" $ do
     let shape       = TF.Shape [1, 3] -- Consider a 3-dim embedding of two items.
-    let embedding1  = [ 1, 1, 1 ] :: [Int32]
-    let embedding2  = [ 0, 0, 0 ] :: [Int32]
+    let embedding1  = [ 1, 1, 1 :: Int32]
```
Comment (blackgnezdo) [inline tensorflow-ops/tests/EmbeddingOpsTest.hs]: Do you feel like running this file through the latest version of hindent? I bet it's no fun to chase down all these space nits (extra one after [)

File: tensorflow-test/src/TensorFlow/Test.hs [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+{-# LANGUAGE OverloadedStrings #-}
+
+module TensorFlow.Test
+    ( assertAllClose
+    ) where
+
+import Test.HUnit.Lang (Assertion(..))
+import qualified Data.Vector as V
+import Test.HUnit ((@?))
```
Comment (blackgnezdo) [inline tensorflow-test/src/TensorFlow/Test.hs]: Sort imports.

**Comment by blackgnezdo** [comment]
Jenkins please test this.

**Comment by blackgnezdo** [comment]
ok to test

File: tensorflow-ops/tests/EmbeddingOpsTest.hs (hunk 13 above)
Comment (silky) [inline tensorflow-ops/tests/EmbeddingOpsTest.hs]: i tried all the available styles and they all missed with all the other indenting in fairly radical ways

i cleaned these up, and a few others. don't mind if you point out things i missed :)

File: tensorflow-ops/src/TensorFlow/Gradient.hs [hunk 15]
```diff
@@ -402,6 +403,21 @@ type GradientFunc a = NodeDef
 toT :: Output -> Tensor Value a
 toT = Tensor ValueKind
 
+
+-- | Wrapper around `TensorFlow.GenOps.Core.slice` that builds vectors from scalars for
+-- simple slicing operations.
+flatSlice :: forall v1 t i . (TensorType t)
+         => Tensor v1 t -- ^ __input__
+         -> Int32 -- ^ __begin__: specifies the offset into the first dimension of
+                            -- 'input' to slice from.
```
Comment (blackgnezdo) [inline tensorflow-ops/src/TensorFlow/Gradient.hs]: Maybe align the -- with the ones above?

File: tensorflow-ops/src/TensorFlow/Ops.hs [hunk 16]
```diff
@@ -256,6 +257,17 @@ constant (Shape shape') values
                       [def & TensorShape.size .~ x | x <- shape']
                 & tensorVal .~ values
 
+-- | Reshape a N-D tensor down to a scalar.
+-- 
+-- See `TensorFlow.GenOps.Core.reshape`.
+scalarize :: (TensorType a) 
+         => Tensor v a 
+         -> Tensor Value a
```
Comment (blackgnezdo) [inline tensorflow-ops/src/TensorFlow/Ops.hs]: Probably the signature will fit on one line?
